app/Models/PostModel.php

```diff
@@ -39,16 +39,107 @@ public function visible()
         return $this->where('visible', true);
     }
 
+    /**
+     * Scope method to only return top level posts.
+     *
+     * Example:
+     *  $posts = model(PostModel::class)->main()->findAll();
+     */
+    public function main()
+    {
+        return $this->where('reply_to', null);
+    }
+
     /**
      * Returns a paginated list of posts for the thread.
      */
     public function forThread(int $threadId, int $perPage = 10)
     {
         $posts = $this->where('thread_id', $threadId)
             ->visible()
+            ->main()
             ->orderBy('id', 'asc')
             ->paginate($perPage);
 
+        if (! $posts) {
+            return [];
+        }
+
+        $posts = $this->withReplies($posts);
+
+        return $this->withUsers($posts);
+    }
+
+    /**
+     * Adds last X replies for every post.
+     */
+    public function withReplies(array $posts, int $limit = 2): array
+    {
+        $postIds      = array_map('intval', array_column($posts, 'id'));
+        $replies      = $this->getReplies($postIds, $limit);
+        $repliesCount = $this->getRepliesCount($postIds);
+
+        foreach ($posts as $post) {
+            $post->replies       = $replies[$post->id] ?? [];
+            $post->replies_count = $repliesCount[$post->id] ?? 0;
+        }
+
+        return $posts;
+    }
+
+    /**
+     * Get latest replies for posts by given IDs.
+     */
+    public function getReplies(array $postIds, int $limit): array
+    {
+        // Used to get last replies using rank
+        $subQuery = $this->builder()
+            ->select('*, ROW_NUMBER() OVER (PARTITION BY reply_to ORDER BY created_at DESC) AS reply_rank')
```

Comment:
This reminds me that I need to go study my SQL some more. I think I could have really used this a couple of days ago at work :)